src/content/blog/rust-based-javascript-linters-lack-typed-linting-right-now/index.mdx

```diff
@@ -0,0 +1,151 @@
+]
+title: "Rust-Based JavaScript Linters Lack Typed Linting Right Now"
+---
+
+One of 2023's biggest trends for web tooling was rewriting existing tooling in [Rust](https://www.rust-lang.org).
+Rust is a wonderful programming language that allows for shockingly fast binaries that still interop well with other web tools courtesy of WebAssembly.
+The speedups seen in tools such as [swc](https://swc.rs) and [Turbopack](https://turbo.build/pack) are very exciting for fast development experiences.
+
+[Biome](https://biomejs.dev), [`deno lint`](https://lint.deno.land), [oxc](https://github.com/oxc-project/oxc), and [RSLint](https://rslint.org) are all projects that include at least a JavaScript/TypeScript linter written in Rust.
+The idea of a linter that runs at Rust (native code) speed rather than JavaScript (JIT script) speed is quite appealing for developers frustrated with slow lint times.
+
+But!
+It's a misconception to think that Rust-based linters are a complete and total replacement for [ESLint](https://eslint.org) today.
+There are always pros and cons to switching tooling.
+In this case, the pros are much less than many have noted, while the cons are much more significant.
+
+## Recap: Type-Checked Linting
+
+Traditionally, lint rules in linters such as ESLint only have visibility into one source code file at a time.
+This makes them fast and theoretically cacheable and parallelizable.
+
+[typescript-eslint](https://typescript-eslint.io) introduces the concept of [linting with type information](https://typescript-eslint.io/linting/typed-linting).
+By calling to TypeScript's type checking APIs, lint rules can make much more informed decisions on code based on types informed by potentially any other file in your project.
+
+Type-checked lint rules can be significantly more capable than traditional lint rules.
+For example:
+
+- [`@typescript-eslint/await-thenable`](https://typescript-eslint.io/rules/await-thenable) disallows `await`s unnecessarily used on non-Thenable (non-Promise) values
+- [`@typescript-eslint/no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises) lets you know if you create a Promise and forget to handle it safely
+- [`@typescript-eslint/no-for-in-array`](https://typescript-eslint.io/rules/no-for-in-array): flags accidental unsafe `for...in` iterations over arrays (instead of `for...of`)
+
+Each of those rules is only practically useful when they can use type information to determine when to report issues.
+Without type information, they wouldn't be able to understand the type of any value imported from another module.
+
+### Type-Checked Linting Performance
+
+The main downside of type-checked linting is performance.
+Lint rules are now blocked on calling to an API such as TypeScript's, which generally need to read all files to see which ones impact types of any other file.
+That means linting performance will often be strictly worse than that of running `tsc` on your entire project.
```

Comment:
Heh I'll remove _"strictly"_ - I suppose it's not an exact superset of runtime behavior, and in theory if you have very few lint rules enabled you might only be calling a few quick type APIs for a small number of nodes. There aren't any benchmarks I have to prove otherwise.

Generally this statement is true in our experience because type checking is _so much slower_ than linting. In general.